src/main/java/com/google/devtools/build/lib/remote/common/RemotePathResolver.java

```diff
@@ -0,0 +1,194 @@
+
+  /**
+   * Resolves the output path relative to input root for the given {@link Path}.
+   */
+  String resolveOutputPath(Path path);
+
+  /**
+   * Resolves the output path relative to input root for the given {@link PathFragment}.
+   *
+   * @param execPath a path fragment relative to {@code execRoot}.
+   */
+  String resolveOutputPath(PathFragment execPath);
+
+  /**
+   * Resolves the output path relative to input root for the {@link ActionInput}.
+   */
+  default String resolveOutputPath(ActionInput actionInput) {
+    return resolveOutputPath(actionInput.getExecPath());
+  }
+
+  /**
+   * Resolves the local {@link Path} of an output file.
+   *
+   * @param outputPath the return value of {@link #resolveOutputPath(PathFragment)}.
+   */
+  Path resolveLocalPath(String outputPath);
+
+  /**
+   * Resolves the local {@link Path} for the {@link ActionInput}.
+   */
+  default Path resolveLocalPath(ActionInput actionInput) {
+    String outputPath = resolveOutputPath(actionInput.getExecPath());
+    return resolveLocalPath(outputPath);
+  }
+
+  /**
+   * Creates the default {@link RemotePathResolver}.
+   */
+  static RemotePathResolver createDefault(Path execRoot) {
+    return new DefaultRemotePathResolver(execRoot);
+  }
+
+  /**
+   * The default {@link RemotePathResolver} which use {@code execRoot} as input root and do NOT
+   * set {@code workingDirectory} for remote actions.
+   */
+  class DefaultRemotePathResolver implements RemotePathResolver {
+
+    final private Path execRoot;
+
+    public DefaultRemotePathResolver(Path execRoot) {
+      this.execRoot = execRoot;
+    }
+
+    @Nullable
+    @Override
+    public String getWorkingDirectory() {
+      return null;
+    }
+
+    @Override
+    public SortedMap<PathFragment, ActionInput> getInputMapping(SpawnExecutionContext context)
+        throws IOException {
+      return context.getInputMapping(PathFragment.EMPTY_FRAGMENT);
+    }
+
+    @Override
+    public String resolveOutputPath(Path path) {
+      return path.relativeTo(execRoot).getPathString();
+    }
+
+    @Override
+    public String resolveOutputPath(PathFragment execPath) {
+      return execPath.getPathString();
+    }
+
+    @Override
+    public Path resolveLocalPath(String outputPath) {
+      return execRoot.getRelative(outputPath);
+    }
+
+    @Override
+    public Path resolveLocalPath(ActionInput actionInput) {
+      return ActionInputHelper.toInputPath(actionInput, execRoot);
+    }
+  }
+
+  /**
+   * A {@link RemotePathResolver} used when {@code --experimental_sibling_external_layout} is set.
+   * Use parent directory of {@code execRoot} and set {@code workingDirectory} to the base name of
+   * {@code execRoot}.
+   *
+   * The paths of outputs are relative to {@code workingDirectory} if
+   * {@code --incompatible_remote_output_paths_relative_to_input_root} is not set, otherwise,
+   * relative to input root.
+   */
+  class SiblingExternalLayoutResolver implements RemotePathResolver {
```

Comment:
nit: `SiblingRepositoryLayout` for consistency with the name of the flag?

Reply:
Done.